split generated VariableType.cpp

anderspapitto opened the pull request:
On my devgpu, this brings the time taken for `touch torch/csrc/jit/type.h && time python setup.py rebuild develop` (debug mode, multicore build) down from 75 seconds to 62 seconds. For the `ninja install` of libtorch portion, which this affects, the reduction is from 52 seconds to 35.

apaszke commented on torch/csrc/autograd/VariableTypeImplementation.h:
```diff
@@ -0,0 +1,182 @@
+      tensors, typename MakeIndices<sizeof...(Tensors)>::indices());
+}
+
+static std::vector<std::vector<int64_t>> to_args_sizes(TensorList tensors) {
```
Those should probably be inline instead of static now.

apaszke commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -1,520 +1,12 @@
-#include "VariableType.h"
+#include "torch/csrc/autograd/VariableTypeImplementation.h"
```
Nit: maybe call this VariableTypeUtils.h? That’s definitely less verbose. We usually use XImpl for the pImpl pattern so it might be confusing.

anderspapitto commented:
the remaining two failures are spurious and/or known existing issues

gchanan reviewed (commented):
I think splitting this up is a good idea, but I think we should attempt to do it in some logical manner (meaning one can reason about what file a given function will end up in).  In particular, one task that I do a lot is to diff the generated code when I make changes; if I do that now, it's possible a ton of functions shift around and I don't even know what file to look in without grepping (adding an extra step each time).

I'm not sure what the best way to logically split this up is, but you can look at Declarations.yaml for  some ideas.
I'd start by splitting via declaration['method_of'] (we already do this somewhat in Type and TypeExtendedInterface), so there is precedence here.  I believe the current possible values are lists of 'namespace', 'Tensor', and 'Type' (the methods that end up here probably all include 'Type').  You could do something like: ('namespace' and not 'Tensor', 'Tensor' and not 'namespace', else).  And see if that makes a reasonable split.

anderspapitto commented:
I've made two changes to account for @gchanan's concerns.

The first change is that we now generate VariableTypeEverything.cpp alongside the various shards. This file is not used by the build itself, but contains all generated definitions. Anyone who wants to see all definitions in one predictable place (for example to compare generated functions between commits) should look here. This is documented in a prominent comment at the top of templates/VariableType.cpp (and therefore also in each generated file).

The second change is that rather than round-robining function definitions, we now sort them alphabetically and take equal-sized contiguous chunks. This will lead to less churn over time (though anyone looking at these generated files manually is still strongly recommended to use VariableTypeEverything.cpp).

@gchanan and I also invested breaking the file up along semantic lines, based on the properties of each decl. However, it's not clear that there is any division which is both semantically pleasing and breaks into equal-sized chunks. For example,

```
In [11]: Counter([tuple([decl['abstract']] + decl['method_of']) for decl in aten_decls])
Out[11]:
Counter({(True, 'Type', 'Tensor'): 90,
         (True, 'Type', 'namespace'): 540,
         (True, 'Type', 'Tensor', 'namespace'): 118,
         (True, 'Type'): 14,
         (False, 'Type', 'namespace'): 344,
         (False, 'Type', 'Tensor', 'namespace'): 152,
         (False, 'Type', 'Tensor'): 28,
         (False, 'Type'): 1})
```

Another downside of a semantic division is that if the distribution were to skew over time as functions are added/removed/modified, it's unlikely that anyone would catch (much less diagnose and fix) the build-perf regression.

anderspapitto committed:
split generated VariableType.cpp (#12493)

Summary:
On my devgpu, this brings the time taken for `touch torch/csrc/jit/type.h && time python setup.py rebuild develop` (debug mode, multicore build) down from 75 seconds to 62 seconds. For the `ninja install` of libtorch portion, which this affects, the reduction is from 52 seconds to 35.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12493

Differential Revision: D10315988

fbshipit-source-id: b3436db28cd507aebe8a11edd7e546158e12a6f1

apaszke commented:
Actually sorting them alphabetically might cause a cascade of a few functions overflowing every single block. It might be a better idea to use a deterministic hash function to split them into approximately equally sized files.

zdevito reviewed (approved):
I think benefit (huge reduction in compile latency) outweighs the downsides of having people who are working on variable type need to look through 5 files to find where a definition is. 

I agree with @apaszke that the decision should be stable: as long as we have 5 bins, an op should always go into the same bin, regardless of the addition or deletion of other ops. This can be done simply by just summing the characters in the name to form a stable hash: 
```
    for a in aten_decls:
        bin = sum(ord(c) for c in a['name']) % 5
```
In practice this is balanced enough `{0: 251, 1: 282, 2: 278, 3: 227, 4: 252}`
